mongo/datadog_checks/mongo/mongo.py

```diff
@@ -158,11 +155,53 @@ def __init__(self, name, init_config, instances=None):
         self.custom_queries = self.instance.get("custom_queries", [])
         # By default consider that this instance is a standalone, updated on each check run.
         self.deployment = StandaloneDeployment()
+        self.collectors = []
+        self.last_states_by_server = {}
 
     @classmethod
     def get_library_versions(cls):
         return {'pymongo': pymongo.version}
 
+    def refresh_collectors(self, mongo_version, all_dbs, tags):
+        collectors = []
+        if isinstance(self.deployment, ReplicaSetDeployment):
+            if self.replica_check:
+                collectors.append(ReplicaCollector(self, tags))
+            collectors.append(ReplicationOpLogCollector(self, tags))
+
+        if not isinstance(self.deployment, MongosDeployment):
+            # The local database contains different information on each node.
+            # But is only available for mongod instances.
+            collectors.append(DbStatCollector(self, "local", tags))
+            collectors.append(FsyncLockCollector(self, self.db_name, tags))
+            if 'top' in self.additional_metrics:
+                collectors.append(TopCollector(self, tags))
+
+        if self.deployment.is_principal():
+            if self.db_name not in all_dbs:
+                raise ConfigurationError(
+                    "Configured database {} does not exist. Available databases are {}.".format(self.db_name, all_dbs)
+                )
+            if 'local' in all_dbs:
+                # Already monitored for all instances
+                all_dbs.remove('local')
+            for db_n in all_dbs:
+                collectors.append(DbStatCollector(self, db_n, tags))
+
+            if self.collections_indexes_stats:
+                if LooseVersion(mongo_version) >= LooseVersion("3.2"):
+                    collectors.append(IndexStatsCollector(self, self.db_name, tags, self.coll_names))
+                else:
+                    msg = "'collections_indexes_stats' is only available starting from mongo 3.2: "
+                    "your mongo version is %s"
+                    self.log.error(msg, mongo_version)
+            collectors.append(CollStatsCollector(self, self.db_name, tags, coll_names=self.coll_names))
+
+        # TODO: Is it the right place for custom queries?
+        collectors.append(CustomQueriesCollector(self, self.db_name, tags, self.custom_queries))
```

Comment:
Tricky stuff here. Collecting custom queries should probably never happen inside a shard as the shard won't have the visibility. For replica set we also likely don't want the same queries to be run multiple times.
But at the same time, if the user did configure some queries, the intent is explicit. Also it can make sense if the query is run on the 'local' database (different in each node)

Opinions welcome.

Reply:
> But at the same time, if the user did configure some queries, the intent is explicit.

…But are users always aware of the caveats listed here?

Mayne we could add flags like `custom_queries_shard_safe: true` and `custom_queries_replica_safe: true` that users should explicitly enable, with some information about caveats of running custom queries depending on the type of the target node. Then we would:

* Check those flags and compare them with the detected deployment.
* If flags and the detected deployment match, then enable the custom queries collector.
* Otherwise, we can still enable the custom queries collector, but also log a warning letting the user know that the custom query may not be safe to run in their case, and that they should either collect those metrics from a different node, or enable <flag> to let the check know it's actually safe.

Reply:
Flags would have to be set to false by default to not break backwards compatibility for existing configuration that make "safe" queries to shards or replica set members. Which reduces drastically their usefulness.

I'm tempted to leave as is with a warning in the configuration. Not running specific custom queries when they are explicitly defined in the configuration would be a bad UX.


I see another drawback with the current approach. For sharding it's not too much of a problem, users can configure the check to collect custom queries only on the "mongos" instance, if they don't they might see incorrect metrics but this is a pilot error at that point.
For replica set it's a bit trickier, mongod node can change from secondary to primary and recursively so there's no way to differentiate upon configuration. That means there is no way to prevent the check from running the same custom query on each node.
So we might consider an alternative situation:

- Break a bit the existing behavior by only running custom queries on a "master" node (not principal, master here being any primary node, any mongos, or any standalone).
- Add a new field in the `custom_queries` config, each query can specify 'run_on_secondaries: true' so that the query is run.

This way, we support all use cases as long as the user configuration is correct.
WDYT?

Reply:
Yes, that sounds legit. :+1: I think given the amount of improvements that will come in the next major of this integration anyway, an additional small breaking change being "custom queries now only run on master nodes by default" sounds acceptable.